pisa-proxy/app/config/src/config.rs

```diff
@@ -21,185 +21,208 @@ use proxy::proxy::{MySQLNode, MySQLNodes, ProxiesConfig, ProxyConfig};
-            .arg(Arg::new("config").short('c').long("config").help("Config path").takes_value(true))
-            .arg(Arg::new("loglevel").long("log-level").help("Log level").takes_value(true))
-            .get_matches();
-        let config: Config;
-
-        let local = match env::var(PISA_PROXY_CONFIG_ENV_LOCAL_CONFIG) {
-            Ok(local) => local,
-            Err(_) => "false".to_string(),
-        };
-
-        if local.eq("true") {
-            let mut config_path = PISA_PROXY_DEFAULT_CONFIG_PATH;
-
-            if let Some(path) = matches.value_of("config") {
-                config_path = path;
-            }
+        .version(PisaProxyConfigBuilder::default().build_from_env().build_version()._version.unwrap().as_str())
+        .arg(Arg::new("port").short('p').long("port").help("Http port").takes_value(true))
+        .arg(Arg::new("config").short('c').long("config").help("Config path").takes_value(true))
+        .arg(Arg::new("loglevel").long("log-level").help("Log level").takes_value(true))
+        .get_matches();
+
+        if let Some(port) = matches.value_of("port") {
+            builder._port = port.to_string();
+        }
+        if let Some(path) = matches.value_of("config") {
+            builder._config_path = path.to_string();
+        }
+        if let Some(loglevel) = matches.value_of("loglevel") {
+            builder._log_level = loglevel.to_string();
+        }
+        builder
+    }
 
-            let mut file = match File::open(config_path) {
-                Err(e) => {
-                    eprintln!("{:?}", e);
-                    std::process::exit(-1);
-                }
-                Ok(file) => file,
-            };
-
-            let mut config_str = String::new();
-            file.read_to_string(&mut config_str).unwrap();
-            config = toml::from_str(&config_str).unwrap();
+    pub fn build_version(mut self) -> Self {
+        if !self._git_tag.is_empty() {
+            self._version = Some(format!("{}", self._git_tag));
         } else {
-            config = PisaConfig::load_http().unwrap();
+            self._version = Some(format!("{}-{}", self._git_branch, self._git_commit));
         }
-        trace!("configs: {:#?}", config);
+        self
+    }
 
-        let mut pisa_config = PisaConfig {
-            admin: config.admin,
-            proxies: vec![],
-            mysql_nodes: vec![],
-            shardingsphere_proxy_nodes: vec![],
-            version: String::default(),
-        };
+    pub fn load_config(mut self) -> PisaProxyConfig {
+        let cmd_builder = PisaProxyConfigBuilder::default().build_from_cmd();
+        let config_path = cmd_builder._config_path.clone();
+        let cmd_config = cmd_builder.build();
 
-        //TODO: need refactor
-        pisa_config.version = PisaConfig::get_version();
+        let env_builder = PisaProxyConfigBuilder::default().build_from_env();
+        let is_local_config = env_builder._local.clone();
```

Comment:
Is it possible to remove `clone`.

Reply:
> Is it possible to remove `clone`.

I’d like to figure this out to see if it will cause a Partial Copy